signature/signer.go

```diff
@@ -0,0 +1,189 @@
+		certs[i], err = x509.ParseCertificate(c)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// create signer
+	return NewSigner(cert.PrivateKey, certs)
+}
+
+// NewSigner creates a signer with the recommended signing method and a signing key bundled
+// with a certificate chain.
+// The relation of the provided signing key and its certificate chain is not verified,
+// and should be verified by the caller.
+func NewSigner(key crypto.PrivateKey, certChain []*x509.Certificate) (notation.Signer, error) {
+	lsp, err := signer.GetLocalSignatureProvider(certChain, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pluginSigner{
+		runner: &builtinPlugin{ localSignatureProvider: lsp },
+	}, nil
+}
+
+// builtinPlugin is a plugin.Runner implementation which
+// signs supports the generate-signature workflow using
+// the provided key and certificates.
+type builtinPlugin struct {
+	localSignatureProvider *signer.LocalSignatureProvider
+	sigAlg                 signer.SignatureAlgorithm
+}
+
+func (builtinPlugin) metadata() *plugin.Metadata {
+	// The only properties that are really relevant
+	// are the supported contract version and the capabilities.
+	// All other are just filled with meaningful data.
+	return &plugin.Metadata{
+		SupportedContractVersions: []string{plugin.ContractVersion},
+		Capabilities:              []plugin.Capability{plugin.CapabilitySignatureGenerator},
+		Name:                      "built-in",
+		Description:               "Notation built-in signer",
+		Version:                   plugin.ContractVersion,
+		URL:                       "https://github.com/notaryproject/notation-go",
+	}
+}
+
+// Run implement the generate-signature workflow.
+func (r *builtinPlugin) Run(_ context.Context, req plugin.Request) (interface{}, error) {
+	switch req.Command() {
+	case plugin.CommandGetMetadata:
+		return r.metadata(), nil
+	case plugin.CommandDescribeKey:
+		req1 := req.(*plugin.DescribeKeyRequest)
+
+		ks, err := r.localSignatureProvider.KeySpec()
+		if err != nil {
+			return nil, plugin.RequestError{
+				Code: plugin.ErrorCodeGeneric,
+				Err:  err,
+			}
+		}
+
+		return &plugin.DescribeKeyResponse{
+			KeyID:   req1.KeyID,
+			KeySpec: ks,
+		}, nil
+	case plugin.CommandGenerateSignature:
+		req1 := req.(*plugin.GenerateSignatureRequest)
+
+		signed, certChain, err := r.localSignatureProvider.Sign(req1.Payload)
+		if err != nil {
+			return nil, plugin.RequestError{
+				Code: plugin.ErrorCodeGeneric,
+				Err:  err,
+			}
+		}
+
+		rawCerts := make([][]byte, len(certChain))
+		for i, cert := range certChain {
+			rawCerts[i] = cert.Raw
+		}
+
+		return &plugin.GenerateSignatureResponse{
+			KeyID:            req1.KeyID,
+			Signature:        signed,
+			SigningAlgorithm: r.sigAlg,
+			CertificateChain: rawCerts,
+		}, nil
+	}
+	return nil, plugin.RequestError{
+		Code: plugin.ErrorCodeGeneric,
+		Err:  fmt.Errorf("command %q is not supported", req.Command()),
+	}
+}
+
+// called from plugin.go
```

Comment:
Remove dead code?

Reply:
It’s the code which contains old time-stamping logic. Technically, it is a dead code but keeping it because it can be used as reference when implementing timestamping in notation-core-go